homeassistant/components/denonavr/config_flow.py

```diff
@@ -0,0 +1,253 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "denonavr"
+
+SUPPORTED_MANUFACTURERS = ["Denon", "DENON", "Marantz"]
+
+CONF_SHOW_ALL_SOURCES = "show_all_sources"
+CONF_ZONE2 = "zone2"
+CONF_ZONE3 = "zone3"
+CONF_TYPE = "type"
+CONF_MODEL = "model"
+CONF_MANUFACTURER = "manufacturer"
+CONF_SERIAL_NUMBER = "serial_number"
+
+DEFAULT_SHOW_SOURCES = False
+DEFAULT_TIMEOUT = 5
+DEFAULT_ZONE2 = False
+DEFAULT_ZONE3 = False
+
+CONFIG_SCHEMA = vol.Schema({vol.Optional(CONF_HOST): str})
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
+    """Options for the component."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry):
+        """Init object."""
+        self.config_entry = config_entry
+
+    async def async_step_init(self, user_input=None):
+        """Manage the options."""
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        settings_schema = vol.Schema(
+            {
+                vol.Optional(
+                    CONF_SHOW_ALL_SOURCES,
+                    default=self.config_entry.options.get(
+                        CONF_SHOW_ALL_SOURCES, DEFAULT_SHOW_SOURCES
+                    ),
+                ): bool,
+                vol.Optional(
+                    CONF_ZONE2,
+                    default=self.config_entry.options.get(CONF_ZONE2, DEFAULT_ZONE2),
+                ): bool,
+                vol.Optional(
+                    CONF_ZONE3,
+                    default=self.config_entry.options.get(CONF_ZONE3, DEFAULT_ZONE3),
+                ): bool,
+            }
+        )
+
+        return self.async_show_form(step_id="init", data_schema=settings_schema)
+
+
+class DenonAvrFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Denon AVR config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the Denon AVR flow."""
+        self.host = None
+        self.serial_number = None
+        self.model_name = None
+        self.timeout = DEFAULT_TIMEOUT
+        self.show_all_sources = DEFAULT_SHOW_SOURCES
+        self.zone2 = DEFAULT_ZONE2
+        self.zone3 = DEFAULT_ZONE3
+        self.d_receivers = []
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry) -> OptionsFlowHandler:
+        """Get the options flow."""
+        return OptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            # check if IP address is set manually
+            host = user_input.get(CONF_HOST)
+            if host:
+                self.host = host
+                return await self.async_step_connect()
+
+            # discovery using denonavr library
+            self.d_receivers = await self.hass.async_add_executor_job(denonavr.discover)
```

Comment:
We need to guard for setting up the same host more than once, since we don't guarantee a unique id.

Reply:
How do you propose to do that?
I thought the unique_id was speciffically ment for that, that was the reason I aborted setup if unique_id could not be set and had a fallback for the unique_id...

Reply:
I could just abort the setup if no serial number could be found?

Reply:
Yes, the config entry unique_id is used to guard from setting up the same id more than once, but since we can't guarantee that a unique id is available we need another way too that guards for this.

Something like this:
```py
for entry in self._async_current_entries():
    if entry.data[CONF_HOST] == host:
        return self.async_abort(reason="already_configured")
```

Reply:
I added it, and also made a test for it.
I used entry.data.get(CONF_HOST) because I think the self._async_current_entries() can also give entities of other integrations right? That do not nessesarly have the CONF_HOST?

Reply:
No, it will only return entries for this domain. Otherwise it would not work for our purposes.

Reply:
alright will make it a key acces then

Reply:
done